src/SharedPipelineComponents/DataLockEvents/DeployDataLock/sql/dml/02 datalock.populate.reference.history.sql

```diff
@@ -1,53 +1,263 @@
+BEGIN
+	CREATE NONCLUSTERED INDEX [DataLockByStatusIdLearnRefNumberPriceEpisodeIdentifier] ON [Reference].[DataLockEvents] ([Status]) INCLUDE (
+		[Id],
+		[LearnRefNumber],
+		[PriceEpisodeIdentifier]
+		)
+END
+/* +++++ start post-op monitoring */
+SET @duration = DATEDIFF(MS, @Now, GETDATE())
+RAISERROR (		@finish,		10,		0,		@operation,		@duration		) WITH NOWAIT
+INSERT INTO [TestStack].[Logs] (	[LogId],	[Logger],	[LogLevel],	[EntryDate],	[Message],	[ErrorDetails]	)
+SELECT NEWID(),	@logsource,	0,	GETDATE(),	@operation + '- Started: [' + @timestamp + '] Duration: [' + cast(@duration AS VARCHAR(255)) + ']',	NULL
+/* ----- finish post-op monitoring*/
+
+
+/* +++++ start pre-op monitoring */
+SELECT @Now = GETDATE()
+SET @operation = 'Identifying latest DataLocks by Provider, Learner and Price Period'
+SET @timestamp = CONVERT(VARCHAR(255), @now, 121)
+RAISERROR (		@start,		10,		0,		@operation,		@timestamp		) WITH NOWAIT
+INSERT INTO [TestStack].[Logs] (	[LogId],	[Logger],	[LogLevel],	[EntryDate],	[Message],	[ErrorDetails]	)
+SELECT NEWID(),	@logsource,	0,	GETDATE(),	@operation + '- Started: [' + @timestamp + '] Duration: [' + cast(@duration AS VARCHAR(255)) + ']',	NULL
+/* ----- finish pre-op monitoring*/
+
+DECLARE @LastestDataLockEvents TABLE ([EventId] BIGINT)
+
 INSERT INTO @LastestDataLockEvents
-SELECT
-	MAX(Id)
-FROM ${DAS_ProviderEvents.FQ}.DataLock.DataLockEvents
-WHERE UKPRN IN (SELECT UKPRN FROM @ProvidersToProcess)
-AND Status <> 3 -- Do not read removed as anything from here will be new again
-GROUP BY UKPRN, LearnRefNumber, PriceEpisodeIdentifier
-
----------------------------------------------------------------
--- DataLockEvents
----------------------------------------------------------------
-INSERT INTO Reference.DataLockEvents
-(Id,DataLockEventId, ProcessDateTime, Status, IlrFileName, SubmittedDateTime, AcademicYear, UKPRN, ULN, LearnRefNumber, AimSeqNumber, 
-PriceEpisodeIdentifier, CommitmentId, EmployerAccountId, EventSource, HasErrors, IlrStartDate, IlrStandardCode, 
-IlrProgrammeType, IlrFrameworkCode, IlrPathwayCode, IlrTrainingPrice, IlrEndpointAssessorPrice, IlrPriceEffectiveFromDate, IlrPriceEffectiveToDate)
-SELECT
-	dle.Id,
-	dle.DataLockEventId,
-	dle.ProcessDateTime, 
-	dle.Status,
-	dle.IlrFileName, 
-    dle.SubmittedDateTime, 
-    dle.AcademicYear, 
-	dle.UKPRN, 
-	dle.ULN, 
-	dle.LearnRefNumber, 
-	dle.AimSeqNumber, 
-	dle.PriceEpisodeIdentifier, 
-	dle.CommitmentId, 
-	dle.EmployerAccountId, 
-	dle.EventSource, 
-	dle.HasErrors, 
-	dle.IlrStartDate, 
-	dle.IlrStandardCode, 
-	dle.IlrProgrammeType, 
-	dle.IlrFrameworkCode, 
-	dle.IlrPathwayCode, 
-	dle.IlrTrainingPrice, 
-	dle.IlrEndpointAssessorPrice,
-	dle.IlrPriceEffectiveFromDate,
-	dle.IlrPriceEffectiveToDate
-	
-FROM ${DAS_ProviderEvents.FQ}.DataLock.DataLockEvents dle
-INNER JOIN @LastestDataLockEvents le
-	ON dle.Id = le.EventId
+SELECT 
+	MAX(dle.[Id]) AS [EventId]
+FROM 
+	[Reference].[DataLockEvents] dle
+INNER JOIN 
+	@ProvidersToProcess ptp
```

Comment:
Is this join necessary?  The query above only pulls events for the providers that have been identified so there wouldn't be a case where there is no record to join to?

Reply:
Good point. Its left over from when I pulled over ALL of the data locks. I will see if it can be removed. I think its probably of minimal physical benefit in elapsed time but from a maintenance perspective is much clearer not to. The query used to take about 7-8 seconds, as this stands it takes about 750 milliseconds. I just tested removing it as suggested and its down to 450 ms.... :-)